### PR TITLE
[backend] Ensure the valid_until date on Indicators is set to a greater value than valid_from when empty (#6836)

### PR DESCRIPTION
**Pull request by SarahBocognano**
### Proposed changes

* According to STIX 2.1 compliance, the until_date must be greater than the valid_from. However right now, if we do not put an until_date manually, we fill the field with the exact same date as valid_from.
* So I add one second to the valid_from if the field until_date is empty

### Related issues

* https://github.com/OpenCTI-Platform/opencti/issues/6836

### Checklist

- [ ] I consider the submitted work as finished
- [ ] I tested the code for its functionality
- [ ] I wrote test cases for the relevant uses case (coverage and e2e)
- [ ] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### CODE REVIEW

**Comment by Archidoit** [comment]
I can have an empty valid_until and a not empty valid_from... 
image

@SarahBocognano : Hmm in which case those this happening ? Because I didn't touch that part, and on my platforme I can have an indicator with no valid_from and no valid_until
image

**Comment by SarahBocognano** [comment]
Update : @Archidoit your point seems valid, need to confirm with product but will probably add the change in the frontend as well to add a automatic fill of valid_until (+1seconde) if valid_from is set

**Comment by SarahBocognano** [comment]
Update: After discussing with the product team, we noticed that there was another issue on this topic. Currently, in the platform, we can set a valid_from equal to valid_until, which is not standard in the STIX format. Therefore, we have decided to add a validation check and an error message to enforce that valid_until must be greater than valid_from to comply with the STIX standard.

To resume : 
- We can have an empty valid_until
- We cannot have a valid_until equal or inferior to a valid_from
- We add one seconde to valid_until if the field was empty AND it's revoked for stix compliance

**Comment by Archidoit** [comment]
I can edit an indicator and have valid_from = valid_until: 
image

**Comment by Archidoit** [comment]
When valid_until = valid_from at indicator creation, the error is badly formatted :
image


--> use Yup to have the same kind of error when valid_from > valid_until 
image

**Comment by Archidoit** [comment]
Indicators with valid_from = valid_until can exist in the platform... Do we have to migrate them and add 1s to valid_until for instance ? @romain-filigran

**Comment by romain-filigran** [comment]
> Indicators with valid_from = valid_until can exist in the platform... Do we have to migrate them and add 1s to valid_until for instance ? @romain-filigran

I don't have a very good view of the difficulty this represents in regard of the potential number of indicators in this situation (particularly due to decay). Do you think it's mandatory ? What is the risk if we don't migrate them, knowing that this should mainly correspond to indicators considered as "Revoked"?

File: opencti-platform/opencti-graphql/src/modules/indicator/indicator-domain.ts [hunk 1]
```diff
@@ -299,12 +299,21 @@ export const addIndicator = async (context: AuthContext, user: AuthUser, indicat
...
+    if (e.key === 'valid_until') [valid_until] = e.value;
+  });
+  if (new Date(valid_until) <= new Date(valid_from)) {
+    throw ValidationError('valid_from', { message: 'The valid until date must be greater than the valid from date' });
+  }
   const scoreEditInput = input.find((e) => e.key === 'x_opencti_score');
   if (scoreEditInput) {
-    const indicator = await findById(context, user, id);
-    if (!indicator) {
-      throw FunctionalError('Cannot edit the field, Indicator cannot be found.');
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/modules/indicator/indicator-domain.ts:306]: why remove this part of code?

**Comment by Archidoit** [comment]
Behavior ok 👍

File: opencti-platform/opencti-graphql/src/modules/indicator/indicator-domain.ts (hunk 1 above)
Comment (SarahBocognano) [reply opencti-platform/opencti-graphql/src/modules/indicator/indicator-domain.ts:306]: I didn't really remove it, I just put it two lines above to be able to use the indicator const not only in the scorEditInput case

**Review by Archidoit — APPROVED** [review]

**Comment by Archidoit** [comment]
Approved, waiting for the PR in client-python to align the behavior

File: opencti-platform/opencti-graphql/src/modules/indicator/indicator-converter.ts [hunk 2]
```diff
@@ -1,9 +1,16 @@
...
 import { STIX_EXT_MITRE, STIX_EXT_OCTI } from '../../types/stix-extensions';
 import type { StixIndicator, StoreEntityIndicator } from './indicator-types';
+import { isNotEmptyField } from '../../database/utils';
 
 const convertIndicatorToStix = (instance: StoreEntityIndicator): StixIndicator => {
   const indicator = buildStixDomain(instance);
+  // Adding one second to the valid_until if valid_from and valid_until are equals,
+  // because according to STIX 2.1 specification the valid_until must be greater than the valid_from.
+  const computedValidUntil = (
+    isNotEmptyField(instance.valid_from) === isNotEmptyField(instance.valid_until)
```
Comment (Archidoit) [inline opencti-platform/opencti-graphql/src/modules/indicator/indicator-converter.ts]:
```suggestion
    isNotEmptyField(instance.valid_from) && isNotEmptyField(instance.valid_until) && instance.valid_until === instance.valid_from
```

Comment (Archidoit) [reply opencti-platform/opencti-graphql/src/modules/indicator/indicator-converter.ts]: isNotEmptyField retourne un boolean, le test suivant retournera toujours vrai si les deux instances sont toutes deux null ou toutes deux non définies:  isNotEmptyField(instance.valid_from) === isNotEmptyField(instance.valid_until)

**Pull request merged**